File: electron/renderer/permission_presets.ts
'use strict';

/* 权限预设（渲染层镜像）。
   执行真值在 app/agent_runtime/permission_presets.py：这里只负责展示与
   选择，桥对未知预设 fail-closed。custom 是派生展示态，不可选。 */

interface PresetOption {
  value: string;
  name: string;
  description: string;
  label: string;
  primary?: boolean;
  confirm?: { title: string; description: string };
}

const PRESETS: PresetOption[] = [
  {
    value: 'plan',
    name: '计划模式',
    label: 'Plan',
    description: 'Create a plan before making changes',
    primary: true,
  },
  {
    value: 'workspace-write',
    name: '接受编辑',
    label: 'Accept edits',
    description: 'Automatically accept all file edits',
    primary: true,
  },
  {
    value: 'danger-full-access',
    name: '完全访问',
    label: 'Bypass permissions',
    description: 'Accepts all permissions',
    primary: true,
    confirm: {
      title: '确认启用 Full access？',
      description:
        '启用 Full access 后，agent 将减少确认步骤，并且可以直接执行更多操作，包括敏感操作、文件修改或外部命令。仅建议在你信任当前任务时使用。',
    },
  },
  {
    value: 'read-only',
    name: '只读',
    label: 'Manual',
    description: 'Always ask before making changes',
    primary: false,
  },
];

const PRIMARY_PRESETS = PRESETS.filter(option => option.primary === true);

const CUSTOM_OPTION: PresetOption = {
  value: 'custom',
  name: '自定义',
  label: '自定义',
  description: '当前权限设置不匹配任何预设。',
};

function optionOf(value: string): PresetOption | undefined {
  if (value === 'custom') return CUSTOM_OPTION;
  return PRESETS.find(option => option.value === value);
}

/** 与 Studio 其余控件一致的 24px / 1.5px 线性图标。 */
function presetSvg(option: PresetOption): string {
  const paths: Record<string, string> = {
    plan: 'M5 6h9M5 11h9M5 16h6M17 5v12M14.5 14.5 17 17l2.5-2.5',
    'read-only': 'M12 3 5 6v5.5c0 4 2.8 7.3 7 8.5 4.2-1.2 7-4.5 7-8.5V6l-7-3Zm-3 8.5 2 2 4-4',
    'workspace-write': 'M3 8a2 2 0 0 1 2-2h3l2 2h9a2 2 0 0 1 2 2v7a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8Zm10.5 7.5 4-4 1.5 1.5-4 4-2 .5.5-2Z',
    'danger-full-access': 'M12 3 5 6v5.5c0 4 2.8 7.3 7 8.5 4.2-1.2 7-4.5 7-8.5V6l-7-3Zm0 5v5m0 3v.1',
    custom: 'M4 7h9m4 0h3M4 12h3m4 0h9M4 17h11m4 0h1M15 5v4M9 10v4m8 1v4',
  };
  return `<svg width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><path d="${paths[option.value] || paths.custom}" /></svg>`;
}

const PermissionPresets = { PRESETS, PRIMARY_PRESETS, optionOf, presetSvg };
if (typeof module !== 'undefined' && module.exports) module.exports = PermissionPresets;
if (typeof globalThis !== 'undefined') {
  (globalThis as typeof globalThis & { PermissionPresets?: typeof PermissionPresets }).PermissionPresets = PermissionPresets;
}

File: app/agent_runtime/permission_presets.py

from __future__ import annotations

from dataclasses import dataclass

from app.agent_runtime.permission_modes import PermissionMode

__all__ = [
    "APPROVAL_POLICIES",
    "CUSTOM_PRESET",
    "PRESETS",
    "SANDBOX_MODES",
    "PermissionPresetSpec",
    "mode_for_preset",
    "preset_select",
    "resolve_preset",
]


SANDBOX_MODES = ("read-only", "workspace-write", "danger-full-access")
APPROVAL_POLICIES = ("ask", "never")

CUSTOM_PRESET = "custom"

CONFIRM_TITLE = "确认启用 Full access？"
CONFIRM_DESCRIPTION = (
    "启用 Full access 后，agent 将减少确认步骤，并且可以直接执行更多操作，"
    "包括敏感操作、文件修改或外部命令。仅建议在你信任当前任务时使用。"
)


@dataclass(frozen=True)
class PermissionPresetSpec:

    sandbox: str
    approval: str
    name: str
    description: str
    confirm: bool = False


PRESETS: dict[str, PermissionPresetSpec] = {
    "auto": PermissionPresetSpec(
        sandbox="workspace-write",
        approval="never",
        name="自动",
        description="可逆的写入直接执行，不可逆的操作和越出工作区的动作仍然会问。",
    ),
    "plan": PermissionPresetSpec(
        sandbox="workspace-write",
        approval="ask",
        name="计划模式",
        description="只读研究并提出计划；用户批准退出计划模式后才能修改。",
    ),
    "read-only": PermissionPresetSpec(
        sandbox="read-only",
        approval="ask",
        name="只读",
        description="只允许读取；任何写入、发送或删除都要先经你确认。",
    ),
    "workspace-write": PermissionPresetSpec(
        sandbox="workspace-write",
        approval="ask",
        name="工作区写入",
        description="工作区内可逆写入直接执行；更大范围的重试需要确认。",
    ),
    "danger-full-access": PermissionPresetSpec(
        sandbox="danger-full-access",
        approval="never",
        name="完全访问",
        description="完整文件访问，不再弹出确认提示。",
        confirm=True,
    ),
}

_PRESET_MODES: dict[str, PermissionMode] = {
    "auto": PermissionMode.ACCEPT_REVERSIBLE,
    "plan": PermissionMode.PLAN,
    "read-only": PermissionMode.SAFE,
    "workspace-write": PermissionMode.DEFAULT,
    "danger-full-access": PermissionMode.BYPASS,
}


def resolve_preset(name: str) -> PermissionPresetSpec:
    return PRESETS[name]


def mode_for_preset(name: str) -> PermissionMode:
    return _PRESET_MODES[name]


def preset_select(current: str) -> dict:
    options = [
        {
            "value": name,
            "name": spec.name,
            "description": spec.description,
            **({"confirmTitle": CONFIRM_TITLE, "confirmDescription": CONFIRM_DESCRIPTION} if spec.confirm else {}),
        }
        for name, spec in PRESETS.items()
    ]
    if current == CUSTOM_PRESET:
        options.append({
            "value": CUSTOM_PRESET,
            "name": "自定义",
            "description": "当前权限设置不匹配任何预设。",
        })
    return {"options": options, "currentValue": current}
